content/dev-tools/xrp-faucets.page.tsx

```diff
@@ -0,0 +1,228 @@
+  setSequence: React.Dispatch<React.SetStateAction<string>>): Promise<void> {
+
+  // Clear existing credentials
+  setGeneratedCredentialsFaucet(selectedFaucet.shortName)
+  setAddress("")
+  setSecret("")
+  setBalance("")
+  setSequence("")
+  const { translate } = useTranslate();
+
+
+  // @ts-expect-error - xrpl is added via a script tag
+  const wallet = xrpl.Wallet.generate()
+  
+  // @ts-expect-error - xrpl is added via a script tag
+  const client = new xrpl.Client(selectedFaucet.wsUrl)
+  await client.connect()
+
+  try {
+
+    setAddress(wallet.address)
+    setSecret(wallet.seed)
+
+    await client.fundWallet(wallet, { faucetHost: selectedFaucet.faucetUrl, usageContext: "xrpl.org-faucet" })
+
+    const response = await waitForSequence(client, wallet.address)
+
+    setSequence(response.sequence)
+    setBalance(response.balance)
+
+  } catch (e) {
+    alert(translate("There was an error with the " + selectedFaucet.shortName + " faucet. Please try again."))
+  }
+}
+
+function TestCredentials({selectedFaucet}) {
+  const { translate } = useTranslate();
+
+  const [generatedCredentialsFaucet, setGeneratedCredentialsFaucet] = useState("")
+  const [address, setAddress] = useState("")
+  const [secret, setSecret] = useState("")
+  const [balance, setBalance] = useState("")
+  const [sequence, setSequence] = useState("")
+
+return (<div>
+    {/* TODO: Once xrpl.js 3.0 is released, replace this with a direct xrpl.js import */}
+    <script src="https://unpkg.com/xrpl@2.5.0-beta.0/build/xrpl-latest-min.js" async />
+
+    {/* <XRPLGuard> TODO: Re-add this once we find a good way to avoid browser/server mismatch errors */}
+      <div className="btn-toolbar" role="toolbar" aria-label="Button"> 
+        <button id="generate-creds-button" onClick={
+            () => generateFaucetCredentialsAndUpdateUI(
+              selectedFaucet, 
+              setGeneratedCredentialsFaucet, 
+              setAddress, 
+              setSecret, 
+              setBalance, 
+              setSequence)
+          } className="btn btn-primary mr-2 mb-2">
+            Generate {selectedFaucet.shortName} credentials
+        </button>
+      </div>
+    {/* </XRPLGuard> */}
+
+    {generatedCredentialsFaucet && <div id="your-credentials">
+      <h2>{translate("Your")} {generatedCredentialsFaucet} {translate("Credentials")}</h2>
+    </div>}
+    
+    {address && <div id="address"><h3>{translate("Address")}</h3>{address}</div>}
+
+    {secret && <div id="secret"><h3>{translate("Secret")}</h3>{secret}</div>}
+    {(address && !balance) && (<div><br/>
+        <div id="loader" style={{display: (address && !balance) ? "inline" : "none"}}>
+          <img alt="(loading)" className="throbber" src="/img/xrp-loader-96.png" /> {translate("Funding account...")}
+        </div>
+      </div>)}
+    
+    {balance && <div id="balance">
+      <h3>{translate("Balance")}</h3>
+      {(Number(balance) * 0.000001).toLocaleString("en")} {translate("XRP")}
```

Comment:
Doesn't `xrpl` have a utility for this calculation?

Reply:
They do, but I can't use `xrpl` at the top-level. Doing so dereferences a null value. It can only be used within functions like `onChange` / `onClick` (I believe because if you include it at the top level it tries to run the code at compile/loading the DOM time instead of after the `<script>` is loaded

Reply:
Redocly added some kind of onready callback function that might be useful for this sort of use case, though I haven't been able to find the docs for how to use it.

Reply:
Using the xrpl one now :) (Beta version works as a direct import)